### PR TITLE
SWIQ-1100 hash password on server as well

### PR DESCRIPTION
**Pull request by rmeissn**
I've implemented password hashing on server side + a migration script to migrate all registered users. This uses the same salt as the one that is used at slidewiki/platform. Is this okay or should it use another salt, that is different?
Please review the migration script carefully!

### CODE REVIEW

**Comment by rmeissn** [comment]
I've forgot to fix the unit-tests. Nevertheless: Is the implementation okay this way?

**Comment by rmeissn** [comment]
I've added @kprist as a reviewer too, as he triggered the discussion about the topic.
@kprist Does this resolve all your mentioned concerns?

**Comment by rmeissn** [comment]
@TBoonX I tried to debug why 18 unit tests are failing: 19 tests on master are failing too. Thus these failing tests were not introduced by my changes. We might need to debug this together ;)

**Comment by kprist** [comment]
It looks good, I can do a test run it if you want, but I'll wait until the unit tests pass. About the testing, may I suggest you use a NODE_ENV=test approach to handling the database so that it doesn't mess with any local database in use? By that I mean something like this commit in deck-service: https://github.com/slidewiki/deck-service/commit/eca6449733906d32534e1bb8cfd40036090231f7

**Review by kprist — COMMENTED** [review]
It looks good, I can do a test run it if you want, but I'll wait until the unit tests pass. About the testing, may I suggest you use a NODE_ENV=test approach to handling the database so that it doesn't mess with any local database in use? By that I mean something like this commit in deck-service: slidewiki/deck-service@eca6449

**Comment by TBoonX** [comment]
@rmeissn the test on master and in your branch are running without error - there is a branch of mine where errors occur but not on these branches. That the test are still running without errors proves that your implementation is correct!
@kprist good suggestion! We should create a branch for this

I still have to test the migration script ...

**Review by TBoonX — APPROVED** [review]
Also the migration script works.
I improved the resetPassword route in a way that clients could provide the used salt. In the future client should also specify the used hash algorithm.

**Comment by TBoonX** [comment]
Hint: there is also a branch on the platform: SWIQ-1100_hashing_password_on_server

**Comment by TBoonX** [comment]
Now the review of @kprist is missing. We should finish this task asap.

File: application/controllers/handler.js [hunk 1]
```diff
@@ -487,7 +488,10 @@ module.exports = {
       }
 
       const newPassword = require('crypto').randomBytes(9).toString('hex');
-      const hashedPassword = JSSHA.sha512(newPassword + config.SMTP.salt);
+      /* The password is hashed one time at the client site (inner hash and optional) and one time at server-side. As we currently only have one salt, it must be the same for slidewiki-platform and the user-service. In case this is splitted, the user-service must know both salts in order to be able to generate a valid password for resetPassword.*/
+      let hashedPassword = co.hashPassword(newPassword, config.SALT);
```
Comment (rmeissn) [inline application/controllers/handler.js:492]: @TBoonX Are you sure you don't need to double hash that one for our official Client? One time for server hashing and another time for client hashing?

Comment (TBoonX) [reply application/controllers/handler.js:492]: Please read the next two lines ;)

**Review by kprist — APPROVED** [review]
Performed the following tests:

1. using a database dump from stable I run deck service and user service locally with the platform
2. tried to log in, and it failed with wrong credentials: expected behaviour as I haven't run the migration script
3. applied the migration script, using the same salt as the one in configuration.js file (fixed a typo to make it work)
4. tried to log in again, and it worked
5. logged out and registered a new user
6. logged in using the new user credentials

File: application/controllers/handler.js (hunk 1 above)
Comment (rmeissn) [reply application/controllers/handler.js:492]: Okay, there's a branch at slidewiki-platform that implements this the other way round. I've overseen that.

**Comment by rmeissn** [comment]
I've tested the reset password code (with branch SWIQ-1100 from slidewiki-platform): works as expected.

I will merge now.

**Pull request merged**